src/util/merkleblock.rs

```diff
@@ -0,0 +1,773 @@
+/// authenticated way.
+///
+/// The encoding works as follows: we traverse the tree in depth-first order,
+/// storing a bit for each traversed node, signifying whether the node is the
+/// parent of at least one matched leaf txid (or a matched txid itself). In
+/// case we are at the leaf level, or this bit is 0, its merkle node hash is
+/// stored, and its children are not explored further. Otherwise, no hash is
+/// stored, but we recurse into both (or the only) child branch. During
+/// decoding, the same depth-first traversal is performed, consuming bits and
+/// hashes as they written during encoding.
+///
+/// The serialization is fixed and provides a hard guarantee about the
+/// encoded size:
+///
+///   SIZE <= 10 + ceil(32.25*N)
+///
+/// Where N represents the number of leaf nodes of the partial tree. N itself
+/// is bounded by:
+///
+///   N <= total_transactions
+///   N <= 1 + matched_transactions*tree_height
+///
+/// The serialization format:
+///  - uint32     total_transactions (4 bytes)
+///  - varint     number of hashes   (1-3 bytes)
+///  - uint256[]  hashes in depth-first order (<= 32*N bytes)
+///  - varint     number of bytes of flag bits (1-3 bytes)
+///  - byte[]     flag bits, packed per 8 in a byte, least significant bit first (<= 2*N-1 bits)
+/// The size constraints follow from this.
+#[derive(PartialEq, Eq, Clone, Debug)]
+pub struct PartialMerkleTree {
+    /// The total number of transactions in the block
+    num_transactions: u32,
+
+    /// node-is-parent-of-matched-txid bits
+    bits: Vec<bool>,
+
+    /// Transaction ids and internal hashes
+    hashes: Vec<sha256d::Hash>,
+}
+
+impl PartialMerkleTree {
+    /// Construct a partial merkle tree
+    /// The `txids` are the transaction hashes of the block and the `matches` is the contains flags
+    /// wherever a tx hash should be included in the proof.
+    ///
+    /// Panics when `txids` is empty or when `matches` has a different length
+    ///
+    /// # Examples
+    ///
+    /// ```ignore
+    /// extern crate bitcoin_hashes;
+    /// use bitcoin_hashes::sha256d;
+    /// use bitcoin_hashes::hex::FromHex;
+    /// use bitcoin::util::merkleblock::PartialMerkleTree;
+    ///
+    /// // Block 80000
+    /// let txids: Vec<sha256d::Hash> = [
+    ///     "c06fbab289f723c6261d3030ddb6be121f7d2508d77862bb1e484f5cd7f92b25",
+    ///     "5a4ebf66822b0b2d56bd9dc64ece0bc38ee7844a23ff1d7320a88c5fdb2ad3e2",
+    /// ]
+    /// .iter()
+    /// .map(|hex| sha256d::Hash::from_hex(hex).unwrap())
+    /// .collect();
+    ///
+    /// // Select the second transaction
+    /// let matches = vec![false, true];
+    /// let tree = PartialMerkleTree::from_txids(&txids, &matches);
+    /// assert!(tree.extract_matches(&mut vec![], &mut vec![]).is_ok());
+    /// ```
+    pub fn from_txids(txids: &[sha256d::Hash], matches: &[bool]) -> Self {
+        // We can never have zero txs in a merkle block, we always need the coinbase tx
+        assert_ne!(txids.len(), 0);
```

Comment:
Don't you have `Error::NoTransactions` for this?

Reply:
In that case we would need to return a `Result<Self, Error>`.

Generally I don't like assert/panics in libraries, except in cases where the input data is incorrect like passing an empty Vec in this case.

I think it's better to leave this panic in place.